Adding first Bazel dif unit test

drewmacrae opened the pull request:
This required a number of supporting changes:
*  Configure bazel to use C++14
*  Add BUILD file for clkmgr dif unit test dependencies
*  Add the clkmgr dif unit test for bazel, can be run with
```bazel test //sw/device/lib/dif:clkmgr_unittest```

Signed-off-by: Drew Macrae <drewmacrae@google.com>

drewmacrae commented:
This is part of # #7972 but I'll start work to add more unit tests at this point.

There are two working unit tests so far. 
```
//sw/device/lib/base/testing:mock_mmio_test                     (cached) PASSED in 0.0s
//sw/device/lib/dif:clkmgr_unittest                             (cached) PASSED in 0.0s
```

drewmacrae commented:
I shortened names for a number of packages when parts of the names were redundant to their paths. I think that's appropriate but I could correct it to longer redundant names if we were concerned about that causing issues in some context.

drewmacrae commented on sw/device/lib/dif/BUILD:
```diff
@@ -0,0 +1,34 @@
+      ],
+  deps = [
+      "//sw/device/lib/base:mmio",
+      "//hw/top_earlgrey/ip/clkmgr/data/autogen:clkmgr_regs",
```
I'm worried this could cause problems if building for englishbreakfast.

cfrantz replied:
I think the way meson currently builds the autogen headers we'd have the same problem.  We can tackle this problem at some later time with a `select` rule for choosing the correct headers.

drewmacrae commented on sw/device/lib/dif/BUILD:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cc_library(
+  name = "clkmgr",
```
this library should be statically linked

cfrantz replied:
It will be when linked into a cc_binary.  For cc_library rules, `linkstatic` determines whether a library is permitted to be statically-linked, dynamically-linked or both.

For `cc_binary` rules, `linkstatic` determines the linking type used when constructing the binary.

IIRC, the default for `cc_binary` is to static link, but dynamic linking is allowed for `cc_test` rules.

drewmacrae commented on sw/device/lib/dif/BUILD:
```diff
@@ -0,0 +1,34 @@
+      ],
+  deps = [
+      ":clkmgr",
+      "//hw/top_earlgrey/ip/clkmgr/data/autogen:clkmgr_regs",
```
If the above dependency on the top_earlgrey clkmgr data is preserved, then this dependency is redundant and uneccesary.

cfrantz replied:
I'm not certain about that.  If `dif_clkmgr_unittest.cc` itself includes `clkmgr_regs.h`, I'm not certain it gets to #include the deps of its dependents.   Might also be a difference in blaze vs bazel.

I usually let `build_cleaner` help me with this sort of thing, but I don't think that tool is available outside of google3.

drewmacrae replied:
It looks like ```build_cleaner``` is only supported for google3 and Blaze, without a plan to support Bazel or external projects. We could mirror lowRisc/OpenTitan in google3, but I'm not sure how far that would get us.

cfrantz replied:
We should not attempt to mirror into google3 - our project environment is too unique and trying to drop the code into g3 just to get `build_cleaner` to work for us is not worth the effort (bazel/blaze differences, toolchain setup, half the project isn't built by `bazel`, coding-style differences, etc).

drewmacrae commented on sw/device/lib/dif/BUILD:
```diff
@@ -0,0 +1,34 @@
+      ":clkmgr",
+      "//hw/top_earlgrey/ip/clkmgr/data/autogen:clkmgr_regs",
+      "@googletest//:gtest_main",
+      "//sw/device/lib/base:mmio",
```
This might also be a redundant include.

cfrantz commented on sw/device/lib/base/BUILD:
```diff
@@ -0,0 +1,50 @@
+        "bitfield.c",
+        "bitfield.h",
+        ],
+    )
+
+cc_library(
+    name = "mem",
+    linkstatic = True,
+    srcs = [
+        "memory.c",
+        "memory.h",
+        ],
+    copts = [
+        "-fno-builtin",
+        ],
+    )
+
+
+cc_library(
+    name = "mmio",
+    linkstatic = True,
+    srcs = [
+        "mmio.c",
+        "mmio.h",
+        ],
+    deps = [
+        ":bitfield",
+        ":mem",
+        ],
+    )
+
+cc_library(
+    name = "hardened",
+    linkstatic = True,
+    srcs = [
+        "hardened.c",
+        "hardened.h",
+        "stdasm.h",
+        ],
+    )
```
You might want to group most of these together into a single target:
https://github.com/cfrantz/opentitan/blob/bazel/sw/device/lib/base/BUILD

Most modules are going to end up using several things from `base` and several of the libraries here are header-only (and we can rely on the linker to do the right thing with individual objects within a library).

cfrantz commented:
You might want to install `buildifier` and run it over your build files to get uniform/correct formatting.

We can also consider (in a future PR) adding a buildifier rule to our codebase to make formatting easier:
https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md

drewmacrae commented on .bazelrc:
```diff
@@ -0,0 +1 @@
+build --cxxopt='-std=c++14'
```
I meant to point to that, Thank you

drewmacrae commented:
The build files in this commit look good in buildifier (I've installed it on my cloudtop but want to play with it a bit before adding it to the repo)

drewmacrae commented on sw/device/lib/dif/BUILD:
```diff
@@ -0,0 +1,34 @@
+      ":dif_warn_unused_result.h",
+      ],
+  deps = [
+      "//sw/device/lib/base:mmio",
```
@cfrantz this target should definitely depend on ...base:mmio as it does here, but I'm running into trouble where I need mmio.h to be built with -DMOCK_MMIO when it's used in a test. The issue appears to be discussed here: https://stackoverflow.com/questions/53950443/build-unit-test-for-c-with-mocked-dependencies-using-bazel#

cfrantz replied:
It appears the DIFs may have been in a less developed state when I built my first bazel experiment.

On thing that I did differently was to separate the interface and link-target for base:mmio (see https://github.com/cfrantz/opentitan/blob/bazel/sw/device/lib/base/BUILD).  I'm not certain that will solve the problem you're pointing to here, as we want "mmio.h" to be preprocessed with `MOCK_MMIO` defined.

One way to get there is to employ the strategy used in the meson rules, which is to include the unit under test into the `srcs` of the test rule rather than it's target into `deps`:

```
cc_library(
  name = "clkmgr",
  srcs = ["dif_clkmgr.c"],
  deps = [ ... ],
)

cc_test(
  name = "clkmgr_unittest",
  srcs = [
    "dif_clkmgr_unittest.cc",
    "dif_clkmgr.c",
    ...
  ],
  deps = [ ... ],
  defines = ["MOCK_MMIO=1"],
)   
```

drewmacrae replied:
That's significantly more elegant than the workaround I had in mind.
```
cc_test(
  name = "gpio_unittest",
  srcs = [
      "dif_gpio_unittest.cc",
      "dif_gpio.c",
      "dif_gpio.h",
      "dif_warn_unused_result.h",
      "//sw/device/lib/base:mmio.c",
      "//sw/device/lib/base:mmio.h",
      ],
  deps = [
      "//hw/ip/gpio/data:gpio_regs",
      "@googletest//:gtest_main",
      "//sw/device/lib/base:mmio",
      "//sw/device/lib/base/testing:mock_mmio",
      ],
  defines = [
      "MOCK_MMIO=1",
      ],
  )
```
appears to work.

cfrantz replied:
You shouldn't need to include "//sw/device/lib/base:mmio.[ch]" into the srcs of the cc_test rule.

In my fork's `bazel` branch, this is handled by `//sw/device/lib/base:base` providing the interface to mmio (ie: `mmio.h`) but not the binary resource.  The binary resource for `mmio` is provided by `//sw/device/lib/base:mmio` and this target is only needed as a final link-time decision for riscv32 cc_binaries.  Similarly, for tests, `//sw/device/lib/base/testing:testing` provides the binary resource for the `mock_mmio` library and satisfies the mmio link-time requirements, thus making it suitable for tests (you can see that `mock_mmio` actually includes the sources for `mmio`, but with `MOCK_MMIO` set, thus fulfilling all the symbol requirements).

The last bit of a sticky point is that the `cc_test` needs define `MOCK_MMIO` so when the srcs `#include "mmio.h"` they have the right preprocessor directives defined (which you already have).

To me, I think that separating interfaces from link-time deps in `//sw/device/lib/base` gives the right amount of flexibility in listing the deps for libraries and tests.  (and `memory.c` and `bitfield.c` were left in the `:base` library target because they don't have any conditional compilation and if they happen to get linked with something that doesn't use them, then the linker/librarian will simply omit them from the target).

Once we start adding support for emitting riscv32 binaries, we can even constrain `mmio` to only riscv32 targets so that it would be impossible to accidentally link it into a test meant to run on x86.

cfrantz commented on sw/device/lib/dif/BUILD:
```diff
@@ -0,0 +1,34 @@
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+cc_library(
+  name = "clkmgr",
+  srcs = [
+      ":dif_clkmgr.c",
+      ":dif_clkmgr.h",
+      ":dif_warn_unused_result.h",
+      ],
```
The `srcs` list should not include the leading colon `:` character.  We're listing files, not providing bazel labels.

The file `dif_clkmgr.h` should be listed in `hdrs` because it is the public interface to dif_clkmgr (if, OTOH, it was a private header file no dependent would be expected to include, _then_ it would belong in `srcs`).

drewmacrae replied:
Oh, I wonder if that fixes anything else. I'll make the fix. I was wondering why bazel distinguished between headers, dependencies and sources.

cfrantz commented on sw/device/lib/base/testing/BUILD:
```diff
@@ -0,0 +1,35 @@
+
+cc_library(
+    name = "mock_mmio",
+    linkstatic = True,
```
I don't think `linkstatic` is necessary here.  This cc_library is effectively testonly and will only get linked with cc_tests which run on the local x86 target rather than the riscv32 target.  Its perfectly fine for this to be subject to `cc_test`'s preferred linking scheme.

drewmacrae replied:
Sounds good, thanks.

cfrantz commented on sw/device/lib/base/testing/BUILD:
```diff
@@ -0,0 +1,35 @@
+    linkstatic = True,
+    srcs = [
+        "mock_mmio.cc",
+        "mock_mmio.h",
+        "mock_mmio_test_utils.h",
```
These should go into `hdrs` as they're the public interface to this library.

drewmacrae commented:
I've addressed the comments here in the other pull request https://github.com/lowRISC/opentitan/pull/8147 So I'll mark that one as not a draft and close this one.
